contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,507 @@
+4. Ordering will probably be necessary on initializers because defaulting in the wild requires ordering. Non-mutating validation on the other hand can be fully parallel.
+5. Some admission depends on knowing the identity of the actor - we will likely need to include the **creator** as information to initializers.
+6. Quota must still run after all validators are invoked. We may need to make quota extensible in the future.
+
+
+#### Initializers
+
+An initializer must allow some API clients to perceive creations prior to other apps. For backwards compatibility,
+uninitialized objects must be invisible to legacy clients. In addition, initializers must be recognized as participating
+in the initialization process and therefore a list of initializers must be populated onto each new object. Like
+finalizers, each initializer should perform an update and remove itself from the object.
+
+Every API object will have the following field added:
+
+```
+type ObjectMeta struct {
+    ...
+    // Initializers is a list of initializers that must run prior to this object being visible to
+    // normal clients. Initializers are executed in order. This field is nil if initialization is not
+    // specified, or empty if no initializers should be run. Only highly privileged users
+    // may modify this field.
+    Initializers []Initializer `json:"initializers"`
+    ...
+}
+
+type Initializer struct {
+    // Name is the name of the process that owns this initializer
+    Name string `json:"name"`
+    // Error is set if the initializer cannot make progress
+    Status *metav1.Status `json:"status"`
+}
+```
+
+On creation, an admission controller defaults the initializers field (if nil) to a value from the system
+configuration that may vary by resource type. If initializers is set to empty slice, or has entries,
+the admission controller will check whether the user has the ability to run the `initialize` verb on
+the current resource type, and reject the entry with 403 if not.
+
+Once created, an object is not visible to clients unless the following conditions are satisfied:
+
+1. The initializers field is an empty array.
+2. The client provides a special option to GET, LIST, or WATCH indicating that the user wishes to see uninitialized objects.
+
+Each initializer is a controller or other client agent that watches for new objects with an initializer
+whose first position matches their assigned name (e.g. `PodAutoSizer`) and then operate on them. These
+clients would use the `?includeUninitializedObjects=true` query param (working name) and observe all
+objects.
+
+The initializer would perform a normal update on the object to perform their function, and then
+remove their entry from `initializers` (or adding more entries). If an error occurs during initialization
+that must terminate initialization, the `Status` field on the initializer should be set instead of removing
+the initializer entry and then the initializer should delete the object.
```

Comment:
Yes, the API server will observe the deletion and return any status found.  Finalizer not required because blocking for initialization is best effort.  If the API server crashes, the client cannot retry their action.